ml-algorithms/src/main/java/org/opensearch/ml/engine/encryptor/EncryptorImpl.java

```diff
@@ -126,131 +130,114 @@ private JceMasterKey createJceMasterKey(String tenantId) {
         return JceMasterKey.getInstance(new SecretKeySpec(bytes, "AES"), "Custom", "", "AES/GCM/NOPADDING");
     }
 
-    private void initMasterKey(String tenantId) {
-        if (tenantMasterKeys.containsKey(Objects.requireNonNullElse(tenantId, DEFAULT_TENANT_ID))) {
+    private void initMasterKey(String tenantId, ActionListener<Boolean> listener) {
+        tenantId = Objects.requireNonNullElse(tenantId, DEFAULT_TENANT_ID);
+        if (tenantMasterKeys.containsKey(tenantId)) {
+            log.info("Master key is already available for tenant {}", tenantId);
+            listener.onResponse(true);
             return;
         }
-        String masterKeyId = MASTER_KEY;
-        if (tenantId != null) {
-            masterKeyId = MASTER_KEY + "_" + hashString(tenantId);
+
+        List<ActionListener<Boolean>> waitingListeners = tenantWaitingListenerMap.computeIfAbsent(tenantId, k -> new ArrayList<>());
+        synchronized (waitingListeners) {
+            if (tenantMasterKeys.containsKey(tenantId)) {
+                log.info("Master key generation is handled by other thread for tenant {}", tenantId);
+                listener.onResponse(true);
+                return;
+            }
+            boolean isFirstThread = waitingListeners.isEmpty();
+            waitingListeners.add(listener);
+            if (!isFirstThread) {
+                log.info("Master key generation is already initiated by other thread for tenant {}", tenantId);
```

Comment:
Should probably add some words indicating that we are queuing the request until the other thread completes.